### ISSUE TITLE
implement the schema decorator

### ISSUE THREAD

**Issue by dimitri-yatsenko** [datajoint/datajoint-python#96]
We had a design meeting and resolved to change the basic mechanism for associating classes with schemas.  Rather than binding python modules to databases, there will be a mechanism to associate an arbitrary context with a given database. 

We will implement the function `dj.schema` to create a decorator that will associate classes with database tables.  Here is an example of the proposed use.

```python
import datajoint as dj

CommonSchema = dj.schema(name='common', context=locals(), connection=dj.conn())

@CommonSchema
class Subject(dj.ManualRelation):
    definition = """
       # basic information about animal subjects used in experiments
       subject_id  : int    # unique animal id
       ----
       species:       enum('mouse', 'rat', 'monkey', 'human', 'dolphin')  # our lab only does these
       date_of_birth: date 
       owner:         varchar(20)  # accountable person
     """

@CommonSchema
class Experiment(dj.ManualRelation):
    definition = """
       # information about an experiment on a given subject
       -> Subject
       experiment_id   :  tinyint   #  experiment number for this subject
       ----
       experiment_date:   date 
       operator:    varchar(20)    # person who did the work
       experiment_notes :  varchar(2000)   
     """
```

This is a major design change that will also involve these internal changes:

1.  The table tier will be indicated by the subclass of `dj.Relation`:  `dj.ManualRelation`, `dj.LookupRelation`, `dj.ImportedRelation`, and `dj.ComputedRelation`.   Users will no longer subclass of `dj.AutoPopulate` since `dj.Imported` and `dj.Computed` will take its place. (How do we indicate if it's a subtable?) 
1.  The table headings and dependencies will no longer be stored centrally by the Connection object but will be managed by each relation object.
1.  The table declaration will only contain the table comment in its first line, prefixed by a pound sign `#`
1.  Foreign key references can be made to any object contained in the context.

These chances will considerably simplify the use of the python version of datajoint by removing some implicit logic of name resolution.

**Comment by dimitri-yatsenko** [comment]
Furthermore, subtables will be indicated by setting the `populate_relation` property to `None`.  If a `dj.ComputedRelation` or a `dj.ImportedRelation` has `populate_relation=None`, objects of this class will display a warning prior to deleting from or drop the table something like "This is a subordinate table and should only be deleted from or dropped together with its main table."

**Comment by eywalker** [comment]
We will go ahead and start implementing this on a separate branch. Till then, we'll keep the current implementation in the `master` branch, but the current implementation should be considered effectively **deprecated** - no additional features will be implemented on the old implementation.

**Comment by eywalker** [comment]
Just to avoid confusion and possible duplicate work - today I'm going to create a branch named `decorator` and start working on implementing decorator factory. I'll commit what I have by the end of the day directly to the branch - I'm also going to comment out all irrelevant test cases. Once the first commit is made onto the `decorator` branch, please continue implementing from there and make pull requests without making direct commits.

**Comment by dimitri-yatsenko** [comment]
This was resolved in 7cb1173f
